Lesson 7/Time/time.c

```diff
@@ -0,0 +1,165 @@
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/proc_fs.h>
+#include <linux/sched.h>
+#include <asm/uaccess.h>
+#include <linux/slab.h>
+
+#include <linux/time.h>
+#include <linux/rtc.h>
+#include <uapi/linux/rtc.h>
+
+
+MODULE_LICENSE("Dual BSD/GPL");
+MODULE_AUTHOR("Dyomin Denis");
+MODULE_DESCRIPTION("Procfs read");
+MODULE_VERSION("0.1");
+
+
+#define MODULE_TAG      "Time module "
+#define PROC_DIRECTORY  "Lesson_7_time"
+#define PROC_FILENAME   "read_time"
+
+
+//static char *proc_buffer;
+struct timespec ts = {0};
+
+static struct proc_dir_entry *proc_dir;
+static struct proc_dir_entry *proc_file;
+
+static int example_read(struct file *file_p, char __user *buffer, size_t length, loff_t *offset);
+
+static struct file_operations proc_fops = {
+    .read  = example_read,
+    //.write = example_write,
+};
+
+/*
+static int create_buffer(void)
+{
+    proc_buffer = kmalloc(BUFFER_SIZE, GFP_KERNEL);
+    if (NULL == proc_buffer)
+        return -ENOMEM;
+    proc_msg_length = 0;
+
+    return 0;
+}
+
+static void cleanup_buffer(void)
+{
+    if (proc_buffer) {
+        kfree(proc_buffer);
+        proc_buffer = NULL;
+    }
+}
+*/
+
+static int create_proc_example(void)
+{
+    proc_dir = proc_mkdir(PROC_DIRECTORY, NULL);
+    if (NULL == proc_dir)
+        return -EFAULT;
+
+    proc_file = proc_create(PROC_FILENAME, S_IFREG | S_IRUGO | S_IWUGO, proc_dir, &proc_fops);
+    if (NULL == proc_file)
+        return -EFAULT;
+
+    return 0;
+}
+
+
+static void cleanup_proc_example(void)
+{
+    if (proc_file)
+    {
+        remove_proc_entry(PROC_FILENAME, proc_dir);
+        proc_file = NULL;
+    }
+    if (proc_dir)
+    {
+        remove_proc_entry(PROC_DIRECTORY, NULL);
+        proc_dir = NULL;
+    }
+}
+
+	
+static int example_read(struct file *file_p, char __user *buffer, size_t length, loff_t *offset)
+{
+
+	//time_t t = time(NULL);
+	
+	struct timeval tv;
+	
+	getnstimeofday(&ts);
+	do_gettimeofday(&tv);
+	//printk("linux time: %d-%d-%d %d:%d:%d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday, tm.tm_hour, tm.tm_min, tm.tm_sec);
+	printk("timespec: %u (sec)\n", ts.tv_sec);
+	printk("timeval: %u (sec)\n", tv.tv_sec);
+
+	//tm = *localtime(&t);
+
+    return 0;
+}
+
+/*
+static int example_write(struct file *file_p, const char __user *buffer, size_t length, loff_t *offset)
+{
+    size_t msg_length;
+    size_t left;
+
+    if (length > BUFFER_SIZE)
+    {
+        printk(KERN_WARNING MODULE_TAG "reduse message length from %u to %u chars\n", length, BUFFER_SIZE);
+        msg_length = BUFFER_SIZE;
+    }
+    else
+        msg_length = length;
+
+    left = raw_copy_from_user(proc_buffer, buffer, msg_length);
+
+    proc_msg_length = msg_length - left;
+    proc_msg_read_pos = 0;
+
+    if (left)
```

Comment:
Dead code again. Let's do a clean up

Reply:
But this lines (with dead code) are commented out.

Reply:
Do you think it's nesessary to merge the code, that was commented out?

Reply:
Sometimes, but I think this is not that case. OK.

Reply:
That's better?